### PR TITLE
Fix S3BlobContainerRetriesTests

### LINKED ISSUE #62390: Failure in org.elasticsearch.repositories.s3.S3BlobContainerRetriesTests.testReadBlobWithRetries
This test just failed in an unrelated PR build https://gradle-enterprise.elastic.co/s/j52pfciigikey/console-log?task=:plugins:repository-s3:test

```
org.elasticsearch.repositories.s3.S3BlobContainerRetriesTests > testReadBlobWithRetries FAILED
    java.lang.AssertionError: Stream [blobName='read_blob_max_retries', range=false] should have reached EOF or should have been aborted but got [eof=false, aborted=false]
        at __randomizedtesting.SeedInfo.seed([681ED21C594165FE:6A2B032834B52D52]:0)
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.assertTrue(Assert.java:41)
        at org.elasticsearch.repositories.s3.S3BlobContainerRetriesTests$AssertingInputStream.close(S3BlobContainerRetriesTests.java:350)
        at org.elasticsearch.common.io.Streams$LimitedInputStream.close(Streams.java:342)
        at org.elasticsearch.core.internal.io.IOUtils.close(IOUtils.java:74)
        at org.elasticsearch.core.internal.io.IOUtils.close(IOUtils.java:116)
        at org.elasticsearch.core.internal.io.IOUtils.close(IOUtils.java:87)
        at org.elasticsearch.core.internal.io.Streams.copy(Streams.java:68)
        at org.elasticsearch.core.internal.io.Streams.copy(Streams.java:91)
        at org.elasticsearch.common.io.Streams.readFully(Streams.java:232)
        at org.elasticsearch.repositories.blobstore.AbstractBlobContainerRetriesTestCase.testReadBlobWithRetries(AbstractBlobContainerRetriesTestCase.java:152)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1750)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:938)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:974)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:988)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:49)
        at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
        at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:48)
        at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
        at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
        at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:817)
        at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:468)
        at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:947)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:832)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:883)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:894)
        at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:41)
        at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
        at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
        at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
        at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
        at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:54)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
        at java.lang.Thread.run(Thread.java:748)

Tests with failures:
 - org.elasticsearch.repositories.s3.S3BlobContainerRetriesTests.testReadBlobWithRetries
```

Could you take a look @tlrx ?

### PR DESCRIPTION
**Pull request by tlrx**
The `AssertingInputStream` in `S3BlobContainerRetriesTests` verifies that `InputStream` are either fully consumed or aborted, but the `eof` flag is only set when the underlying stream returns it.

When buffered read are executed and when the exact number of remaining bytes are read, the `eof` flag is not set to `true`. Instead the test should rely on the total number of bytes read to know if the stream has been fully consumed.

Close #62390

### CODE REVIEW

File: plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java [hunk 1]
```diff
@@ -231,7 +231,7 @@ public void reset() {
 
     // package-private for tests
     boolean isEof() {
-        return eof;
+        return eof | start + currentOffset == currentStreamLastOffset;
```
Comment (original-brownbear) [inline plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java:234]: Probably better to stick with `||` out of principle even if the right hand side is cheap here :)

**Review by original-brownbear — APPROVED** [review]
LGTM, maybe fix the conditional though :)

**Comment by tlrx** [comment]
@elasticmachine update branch

**Comment by tlrx** [comment]
Thanks Armin

**Pull request merged**